File: src/main/java/dev/paigewatson/layoutmaster/data/RollingStockMongoDAL.java
package dev.paigewatson.layoutmaster.data;

import dev.paigewatson.layoutmaster.models.rollingstock.AARDesignation;
import dev.paigewatson.layoutmaster.models.rollingstock.FreightCar;
import dev.paigewatson.layoutmaster.models.rollingstock.RollingStock;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.query.Query;

import java.util.Collections;
import java.util.List;

import static org.springframework.data.mongodb.core.query.Criteria.where;
import static org.springframework.data.mongodb.core.query.Query.query;

public class RollingStockMongoDAL implements RollingStockDAL
{
    private final MongoTemplate mongoTemplate;
    private final String collectionName = "FreightCars";

    public RollingStockMongoDAL(@Autowired MongoTemplate mongoTemplate)
    {
        this.mongoTemplate = mongoTemplate;
    }

    @Override
    public List<RollingStock> getAllRollingStock()
    {
        final List<FreightCar> allReturnedFreightCars = mongoTemplate.findAll(FreightCar.class, collectionName);
        return Collections.unmodifiableList(allReturnedFreightCars);
    }

    @Override
    public void deleteAll()
    {
        mongoTemplate.remove(new Query(), collectionName);
    }

    @Override
    public void delete(RollingStock rollingStockToDelete)
    {
        mongoTemplate.remove(rollingStockToDelete, collectionName);
    }

    @Override
    public List<RollingStock> getAllOfAARDesignation(AARDesignation aarDesignation)
    {
        final List<FreightCar> freightCarsByAARList = mongoTemplate.find(
                query(
                        where("carType.aarDesignation").is(aarDesignation.name())
                ),
                FreightCar.class,
                collectionName
        );
        return Collections.unmodifiableList(freightCarsByAARList);
    }

    @Override
    public List<RollingStock> getAllOfRoadName(String roadName)
    {
        final List<FreightCar> freightCarList = mongoTemplate.find(
                query(
                        where("roadName").is(roadName)
                ),
                FreightCar.class,
                collectionName
        );
        return Collections.unmodifiableList(freightCarList);
    }

    @Override
    public RollingStock insertRollingStock(RollingStock rollingStockToSave)
    {
        return mongoTemplate.insert(rollingStockToSave, collectionName);
    }
}

File: src/main/kotlin/dev/paigewatson/layoutmaster/data/RollingStockMongoDAL.kt
package dev.paigewatson.layoutmaster.data

import dev.paigewatson.layoutmaster.models.rollingstock.AARDesignation
import dev.paigewatson.layoutmaster.models.rollingstock.FreightCar
import dev.paigewatson.layoutmaster.models.rollingstock.RollingStock
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.mongodb.core.MongoTemplate
import org.springframework.data.mongodb.core.query.Criteria
import org.springframework.data.mongodb.core.query.Query
import org.springframework.stereotype.Service
import java.util.*

@Service
class RollingStockMongoDAL(@param:Autowired private val mongoTemplate: MongoTemplate) : RollingStockDAL {
    private val collectionName = "FreightCars"
    override val allRollingStock: List<RollingStock>
        get() {
            val allReturnedFreightCars = mongoTemplate.findAll(FreightCar::class.java, collectionName)
            return Collections.unmodifiableList(allReturnedFreightCars)
        }

    override fun deleteAll() {
        mongoTemplate.remove(Query(), collectionName)
    }

    override fun delete(rollingStockToDelete: RollingStock) {
        mongoTemplate.remove(rollingStockToDelete, collectionName)
    }

    override fun getAllOfAARDesignation(aarDesignation: AARDesignation): List<RollingStock> {
        val freightCarsByAARList = mongoTemplate.find(
            Query.query(
                Criteria.where("carType.aarDesignation").`is`(aarDesignation.name)
            ),
            FreightCar::class.java,
            collectionName
        )
        return Collections.unmodifiableList(freightCarsByAARList)
    }

    override fun getAllOfRoadName(roadName: String): List<RollingStock> {
        val freightCarList = mongoTemplate.find(
            Query.query(
                Criteria.where("roadName").`is`(roadName)
            ),
            FreightCar::class.java,
            collectionName
        )
        return Collections.unmodifiableList(freightCarList)
    }

    override fun insertRollingStock(rollingStockToSave: RollingStock): RollingStock {
        return mongoTemplate.insert(rollingStockToSave, collectionName)
    }
}
